mycpp/gc_dict.h

```diff
@@ -307,49 +278,115 @@ void Dict<K, V>::clear() {
   len_ = 0;
 }
 
-// Returns the position in the array.  Used by dict_contains(), index(),
-// get(), and set().
-//
-// For now this does a linear search.
 // TODO:
-// - hash functions, and linear probing.
-// - resizing based on load factor
-//   - which requires rehashing (re-insert all items)
 // - Special case to intern Str* when it's hashed?  How?
 //   - Should we have wrappers like:
 //   - V GetAndIntern<V>(D, &string_key)
 //   - SetAndIntern<V>(D, &string_key, value)
 //   This will enable duplicate copies of the string to be garbage collected
 template <typename K, typename V>
-int Dict<K, V>::position_of_key(K key) {
-  for (int i = 0; i < capacity_; ++i) {
-    int special = entry_->items_[i];  // NOT an index now
-    if (special == kDeletedEntry) {
-      continue;  // keep searching
+int Dict<K, V>::hash_and_probe(K key) const {
+  if (capacity_ == 0) {
+    return kNotFound;
+  }
+
+  // Hash the key onto a slot in the index. If the first slot is occupied, probe
+  // until an empty one is found.
+  unsigned h = hash_key(key);
+  int init_bucket = h % capacity_;
+
+  // If there's a vacancy because of deletion on the probe path for this key, we
+  // should fill it before consuming a slot that has never been used. This
+  // should help keep the index somewhat compact.
+  int tombstone = kNotFound;
```

Comment:
hm why is this var called "tombstone" ?  It's the return value?

I think of kDeletedEntry as the tombstone, so this reads funny

the tombstone meaning "this position in index_ used to point to some value in keys_ and values_, but it was deleted"

Reply:
I dunno. It reads fine to me... It's not the return value all the time, like in this snippet from a few lines down:
```c++
      if (tombstone != kNotFound) {
        return tombstone;
      }
      return i;
```

Not really sure what name would be better without being verbose

Reply:
OK sounds good, yeah I didn't understand what this code was doing